spring-integration-stomp/src/main/java/org/springframework/integration/stomp/AbstractStompSessionManager.java

```diff
@@ -0,0 +1,204 @@
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.integration.stomp.event.StompExceptionEvent;
+import org.springframework.messaging.simp.stomp.StompCommand;
+import org.springframework.messaging.simp.stomp.StompHeaders;
+import org.springframework.messaging.simp.stomp.StompSession;
+import org.springframework.messaging.simp.stomp.StompSessionHandler;
+import org.springframework.messaging.simp.stomp.StompSessionHandlerAdapter;
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.util.concurrent.ListenableFutureCallback;
+
+/**
+ * @author Artem Bilan
+ * @since 4.2
+ */
+public abstract class AbstractStompSessionManager implements StompSessionManager, ApplicationEventPublisherAware,
+		DisposableBean {
+
+	protected final Log logger = LogFactory.getLog(getClass());
+
+	private final CompositeStompSessionHandler compositeStompSessionHandler = new CompositeStompSessionHandler();
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	private volatile StompHeaders connectHeaders;
+
+	private volatile ListenableFuture<StompSession> stompSessionListenableFuture;
+
+	private volatile StompSession stompSession;
+
+	private volatile boolean autoReceipt;
+
+	private volatile long destroyTimeout = 3000L;
+
+	public void setAutoReceipt(boolean autoReceipt) {
+		this.autoReceipt = autoReceipt;
+	}
+
+	@Override
+	public boolean isAutoReceiptEnabled() {
+		return this.autoReceipt;
+	}
+
+	public void setDestroyTimeout(long destroyTimeout) {
+		this.destroyTimeout = destroyTimeout;
+	}
+
+	public void setConnectHeaders(StompHeaders connectHeaders) {
+		this.connectHeaders = connectHeaders;
+	}
+
+	protected StompHeaders getConnectHeaders() {
+		return connectHeaders;
+	}
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
+
+	@Override
+	public synchronized void connect(StompSessionHandler handler) {
+		this.compositeStompSessionHandler.addHandler(handler);
+		if (this.stompSessionListenableFuture == null) {
+			this.stompSessionListenableFuture = doConnect(this.compositeStompSessionHandler);
+			this.stompSessionListenableFuture.addCallback(new ListenableFutureCallback<StompSession>() {
+
+				@Override
+				public void onFailure(Throwable e) {
+					logger.error("STOMP connect error.", e);
+					if (applicationEventPublisher != null) {
+						applicationEventPublisher.publishEvent(
+								new StompExceptionEvent(AbstractStompSessionManager.this, e));
+					}
+					AbstractStompSessionManager.this.stompSession = null;
+					AbstractStompSessionManager.this.stompSessionListenableFuture = null;
```

Comment:
This looks a bit smelly - won't deal with concurrent `connect` calls and the `start()` methods have no protection against that.